specification/under_development/Compilation_And_Targeting.md

```diff
@@ -0,0 +1,36 @@
+# Compilation Stages
+
+Requirements:
+- Don't fragment the ecosystem unless there is a clear reason why the split makes sense (not sure about a comprehensive positive phrasing)
+- Allow to make use of backend specific instructions at the end user level (e.g. for experimentation and research purposes) without requiring frontends or other backends to add support on a case-by-case basis
+- Define clear guideance for defining backend specific instruction sets
+
+Compilation stages
+- Language specific compilation: Compilation of source code to QIR bitcode
+- General optimizations: Optimizations that map QIR -> QIR and that do not limit compatibility in any way
+- General linking: Combines bitcode from different sources and libraries according to [standard linking practices]()
+- QIS resolution: Links a library (bitcode) taking naming conventions into account and giving strict precedence to the definitions in that library over definitions in the source code, unless otherwise indicated by a function attribute, and resolves precision if needed.
+- Targeting and profile/qis specific optimization: maps QIR -> QIR profile
+- Profile Validation: checks whether the compiled code is compliant with the specified profile and fails compilation otherwise
+- Backend specific compilation and optimization, resolution of runtime functions (possibly object level linking, standard LTO), possibly machine code generation
+
+TODO: "Automatic weak attribute" for `__quantum__qis__` functions?  
+-> probably better to just require that `__quantum__qis__` function either are declaration only, or are weak definitions. 
+-> Or allow for front end specific names with an attribute marking the compiled name that follows naming convention?
```

Comment:
I think we actually want `linkonce` since that allows uncalled definitions to be discarded. But this depends on whether or not we want to maintain uncalled definitions such that later optimizations or rewrite steps can utilize qis functions that are not used by the program itself.